File: devtools/ide/1.3.0/Phalcon/Mvc/Model/Query/Builder/Select.php
<?php 

namespace Phalcon\Mvc\Model\Query\Builder {

	/**
	 * Phalcon\Mvc\Model\Query\Builder\Select
	 *
	 *<code>
	 *$resultset = $this->modelsManager->createBuilder()
	 *   ->from('Robots')
	 *   ->join('RobotsParts')
	 *   ->limit(20)
	 *   ->orderBy('Robots.name')
	 *   ->getQuery()
	 *   ->execute();
	 *</code>
	 */
	
	class Select extends \Phalcon\Mvc\Model\Query\Builder\Join implements \Phalcon\Mvc\Model\Query\BuilderInterface, \Phalcon\Di\InjectionAwareInterface, \Phalcon\Events\EventsAwareInterface {

		protected $_type;

		protected $_columns;

		protected $_models;

		protected $_group;

		protected $_having;

		protected $_order;

		protected $_limit;

		protected $_offset;

		protected $_forUpdate;

		protected $_sharedLock;

		protected $_distinct;

		/**
		 * \Phalcon\Mvc\Model\Query\Builder\Select constructor
		 *
		 *<code>
		 * $params = array(
		 *    'models'     => array('Users'),
		 *    'columns'    => array('id', 'name', 'status'),
		 *    'conditions' => array(
		 *        array(
		 *            "created > :min: AND created < :max:",
		 *            array("min" => '2013-01-01',   'max' => '2014-01-01'),
		 *            array("min" => PDO::PARAM_STR, 'max' => PDO::PARAM_STR),
		 *        ),
		 *    ),
		 *    // or 'conditions' => "created > '2013-01-01' AND created < '2014-01-01'",
		 *    'group'      => array('id', 'name'),
		 *    'having'     => "name = 'Kamil'",
		 *    'order'      => array('name', 'id'),
		 *    'limit'      => 20,
		 *    'offset'     => 20,
		 *    // or 'limit' => array(20, 20),
		 *);
		 *$queryBuilder = new \Phalcon\Mvc\Model\Query\Builder\Select($params);
		 *</code>
		 *
		 * @param array $params
		 * @param \Phalcon\Di $dependencyInjector
		 */
		public function __construct($params=null){ }


		/**
		 * Sets SELECT DISTINCT / SELECT ALL flag
		 *
		 * @param bool|null distinct
		 * @return \Phalcon\Mvc\Model\Query\Builder\Select
		 */
		public function distinct($distinct){ }


		/**
		 * Returns SELECT DISTINCT / SELECT ALL flag
		 *
		 * @return boolean
		 */
		public function getDistinct(){ }


		/**
		 * Sets the columns to be queried
		 *
		 *<code>
		 *	$builder->columns(array('id', 'name'));
		 *</code>
		 *
		 * @param string|array $columns
		 * @return \Phalcon\Mvc\Model\Query\Builder\Select
		 */
		public function columns($columns){ }


		/**
		 * Return the columns to be queried
		 *
		 * @return string|array
		 */
		public function getColumns(){ }


		/**
		 * Sets the models who makes part of the query
		 *
		 *<code>
		 *	$builder->from('Robots');
		 *	$builder->from(array('Robots', 'RobotsParts'));
		 *</code>
		 *
		 * @param string|array $model
		 * @return \Phalcon\Mvc\Model\Query\Builder\Select
		 */
		public function from($model, $alias=null, $merge=null){ }


		/**
		 * Add a model to take part of the query
		 *
		 *<code>
		 *	$builder->addFrom('Robots', 'r');
		 *</code>
		 *
		 * @param string|array $model
		 * @param string $alias
		 * @return \Phalcon\Mvc\Model\Query\Builder\Select
		 */
		public function addFrom($model, $alias=null){ }


		/**
		 * Return the models who makes part of the query
		 *
		 * @return array
		 */
		public function getFrom(){ }


		/**
		 * Sets a ORDER BY condition clause
		 *
		 *<code>
		 *	$builder->orderBy('Robots.name');
		 *	$builder->orderBy(array('1', 'Robots.name'));
		 *</code>
		 *
		 * @param string|array $orderBy
		 * @return \Phalcon\Mvc\Model\Query\Builder\Select
		 */
		public function orderBy($orderBy){ }


		/**
		 * Returns the set ORDER BY clause
		 *
		 * @return string|array
		 */
		public function getOrderBy(){ }


		/**
		 * Sets a HAVING condition clause. You need to escape PHQL reserved words using [ and ] delimiters
		 *
		 *<code>
		 *	$builder->having('SUM(Robots.price) > 0');
		 *</code>
		 *
		 * @param string $having
		 * @return \Phalcon\Mvc\Model\Query\Builder\Select
		 */
		public function having($having){ }


		/**
		 * Return the current having clause
		 *
		 * @return string|array
		 */
		public function getHaving(){ }


		/**
		 * Sets a LIMIT clause, optionally a offset clause
		 *
		 *<code>
		 *	$builder->limit(100);
		 *	$builder->limit(100, 20);
		 *</code>
		 *
		 * @param int $limit
		 * @param int $offset
		 * @return \Phalcon\Mvc\Model\Query\Builder\Select
		 */
		public function limit($limit, $offset=null){ }


		/**
		 * Returns the current LIMIT clause
		 *
		 * @return string|array
		 */
		public function getLimit(){ }


		/**
		 * Sets an OFFSET clause
		 *
		 *<code>
		 *	$builder->offset(30);
		 *</code>
		 *
		 * @param int $limit
		 * @param int $offset
		 * @return \Phalcon\Mvc\Model\Query\Builder\Select
		 */
		public function offset($offset){ }


		/**
		 * Returns the current OFFSET clause
		 *
		 * @return string|array
		 */
		public function getOffset(){ }


		/**
		 * Sets a GROUP BY clause
		 *
		 *<code>
		 *	$builder->groupBy(array('Robots.name'));
		 *</code>
		 *
		 * @param string|array $group
		 * @return \Phalcon\Mvc\Model\Query\Builder\Select
		 */
		public function groupBy($group){ }


		/**
		 * Returns the GROUP BY clause
		 *
		 * @return string|array
		 */
		public function getGroupBy(){ }


		/**
		 * Returns a PHQL statement built based on the builder parameters
		 *
		 * @return string
		 */
		protected function _compile(){ }

	}
}

File: ext/db/builder/select.c
/*
  +------------------------------------------------------------------------+
  | Phalcon Framework                                                      |
  +------------------------------------------------------------------------+
  | Copyright (c) 2011-2014 Phalcon Team (http://www.phalconphp.com)       |
  +------------------------------------------------------------------------+
  | This source file is subject to the New BSD License that is bundled     |
  | with this package in the file docs/LICENSE.txt.                        |
  |                                                                        |
  | If you did not receive a copy of the license and are unable to         |
  | obtain it through the world-wide-web, please send an email             |
  | to license@phalconphp.com so we can send you a copy immediately.       |
  +------------------------------------------------------------------------+
  | Authors: Andres Gutierrez <andres@phalconphp.com>                      |
  |          Eduar Carvajal <eduar@phalconphp.com>                         |
  |          ZhuZongXin <dreamsxin@qq.com>                                 |
  +------------------------------------------------------------------------+
*/

#include "db/builder/select.h"
#include "db/builder/join.h"
#include "db/builder/exception.h"
#include "db/adapterinterface.h"

#include "kernel/main.h"
#include "kernel/memory.h"
#include "kernel/array.h"
#include "kernel/object.h"
#include "kernel/exception.h"
#include "kernel/fcall.h"
#include "kernel/concat.h"
#include "kernel/operators.h"
#include "kernel/string.h"
#include "kernel/file.h"
#include "kernel/debug.h"

/**
 * Phalcon\Db\Builder\Select
 */
zend_class_entry *phalcon_db_builder_select_ce;

PHP_METHOD(Phalcon_Db_Builder_Select, __construct);
PHP_METHOD(Phalcon_Db_Builder_Select, distinct);
PHP_METHOD(Phalcon_Db_Builder_Select, columns);
PHP_METHOD(Phalcon_Db_Builder_Select, orderBy);
PHP_METHOD(Phalcon_Db_Builder_Select, having);
PHP_METHOD(Phalcon_Db_Builder_Select, limit);
PHP_METHOD(Phalcon_Db_Builder_Select, offset);
PHP_METHOD(Phalcon_Db_Builder_Select, groupBy);
PHP_METHOD(Phalcon_Db_Builder_Select, _execute);
PHP_METHOD(Phalcon_Db_Builder_Select, count);

ZEND_BEGIN_ARG_INFO_EX(arginfo_phalcon_db_builder_select___construct, 0, 0, 1)
	ZEND_ARG_INFO(0, tables)
	ZEND_ARG_TYPE_INFO(0, db, IS_STRING, 1)
ZEND_END_ARG_INFO()

ZEND_BEGIN_ARG_INFO_EX(arginfo_phalcon_db_builder_select_distinct, 0, 0, 1)
	ZEND_ARG_TYPE_INFO(0, distinct, _IS_BOOL, 1)
ZEND_END_ARG_INFO()

ZEND_BEGIN_ARG_INFO_EX(arginfo_phalcon_db_builder_select_columns, 0, 0, 1)
	ZEND_ARG_INFO(0, columns)
ZEND_END_ARG_INFO()

ZEND_BEGIN_ARG_INFO_EX(arginfo_phalcon_db_builder_select_orderby, 0, 0, 1)
	ZEND_ARG_TYPE_INFO(0, orderBy, IS_STRING, 1)
ZEND_END_ARG_INFO()

ZEND_BEGIN_ARG_INFO_EX(arginfo_phalcon_db_builder_select_having, 0, 0, 1)
	ZEND_ARG_INFO(0, having)
ZEND_END_ARG_INFO()

ZEND_BEGIN_ARG_INFO_EX(arginfo_phalcon_db_builder_select_limit, 0, 0, 1)
	ZEND_ARG_TYPE_INFO(0, limit, IS_LONG, 0)
	ZEND_ARG_TYPE_INFO(0, offset, IS_LONG, 1)
ZEND_END_ARG_INFO()

ZEND_BEGIN_ARG_INFO_EX(arginfo_phalcon_db_builder_select_offset, 0, 0, 1)
	ZEND_ARG_TYPE_INFO(0, offset, IS_LONG, 1)
ZEND_END_ARG_INFO()

ZEND_BEGIN_ARG_INFO_EX(arginfo_phalcon_db_builder_select_groupby, 0, 0, 1)
	ZEND_ARG_INFO(0, group)
ZEND_END_ARG_INFO()

static const zend_function_entry phalcon_db_builder_select_method_entry[] = {
	PHP_ME(Phalcon_Db_Builder_Select, __construct, arginfo_phalcon_db_builder_select___construct, ZEND_ACC_PUBLIC|ZEND_ACC_CTOR)
	PHP_ME(Phalcon_Db_Builder_Select, distinct, arginfo_phalcon_db_builder_select_distinct, ZEND_ACC_PUBLIC)
	PHP_ME(Phalcon_Db_Builder_Select, columns, arginfo_phalcon_db_builder_select_columns, ZEND_ACC_PUBLIC)
	PHP_ME(Phalcon_Db_Builder_Select, orderBy, arginfo_phalcon_db_builder_select_orderby, ZEND_ACC_PUBLIC)
	PHP_ME(Phalcon_Db_Builder_Select, having, arginfo_phalcon_db_builder_select_having, ZEND_ACC_PUBLIC)
	PHP_ME(Phalcon_Db_Builder_Select, limit, arginfo_phalcon_db_builder_select_limit, ZEND_ACC_PUBLIC)
	PHP_ME(Phalcon_Db_Builder_Select, offset, arginfo_phalcon_db_builder_select_offset, ZEND_ACC_PUBLIC)
	PHP_ME(Phalcon_Db_Builder_Select, groupBy, arginfo_phalcon_db_builder_select_groupby, ZEND_ACC_PUBLIC)
	PHP_ME(Phalcon_Db_Builder_Select, _execute, NULL, ZEND_ACC_PROTECTED)
	PHP_ME(Phalcon_Db_Builder_Select, count, NULL, ZEND_ACC_PUBLIC)
	PHP_FE_END
};

/**
 * Phalcon\Db\Builder\Select initializer
 */
PHALCON_INIT_CLASS(Phalcon_Db_Builder_Select){

	PHALCON_REGISTER_CLASS_EX(Phalcon\\Db\\Builder, Select, db_builder_select, phalcon_db_builder_join_ce, phalcon_db_builder_select_method_entry, 0);

	return SUCCESS;
}

/**
 * Phalcon\Db\Builder\Select constructor
 *
 * @param string|array $tables
 * @param string $db
 */
PHP_METHOD(Phalcon_Db_Builder_Select, __construct){

	zval *tables, *db = NULL, columns = {};

	phalcon_fetch_params(0, 1, 1, &tables, &db);

	if (Z_TYPE_P(tables) != IS_STRING && Z_TYPE_P(tables) != IS_ARRAY) {
		PHALCON_THROW_EXCEPTION_STR(phalcon_db_builder_exception_ce, "The tables must be string or array");
		return;
	}

	phalcon_update_property_array_str(getThis(), SL("_definition"), SL("tables"), tables);

	if (db && PHALCON_IS_NOT_EMPTY(db)) {
		phalcon_update_property(getThis(), SL("_defaultConnectionService"), db);
	}

	ZVAL_STRINGL(&columns, "*", 1);
	phalcon_update_property_array_str(getThis(), SL("_definition"), SL("columns"), &columns);
	zval_ptr_dtor(&columns);
}

/**
 * Sets SELECT DISTINCT / SELECT ALL flag
 *
 * @param bool|null distinct
 * @return Phalcon\Db\Builder\Select
 */
PHP_METHOD(Phalcon_Db_Builder_Select, distinct){

	zval *distinct;

	phalcon_fetch_params(0, 1, 0, &distinct);

	phalcon_update_property_array_str(getThis(), SL("_definition"), SL("distinct"), distinct);
	RETURN_THIS();
}

/**
 * Sets the columns to be queried
 *
 * @param string|array $columns
 * @return Phalcon\Db\Builder\Select
 */
PHP_METHOD(Phalcon_Db_Builder_Select, columns){

	zval *columns;

	phalcon_fetch_params(0, 1, 0, &columns);

	phalcon_update_property_array_str(getThis(), SL("_definition"), SL("columns"), columns);
	RETURN_THIS();
}

/**
 * Sets a ORDER BY condition clause
 *
 * @param string|array $orderBy
 * @return Phalcon\Db\Builder\Select
 */
PHP_METHOD(Phalcon_Db_Builder_Select, orderBy){

	zval *order_by;

	phalcon_fetch_params(0, 1, 0, &order_by);

	phalcon_update_property_array_str(getThis(), SL("_definition"), SL("order"), order_by);
	RETURN_THIS();
}

/**
 * Sets a HAVING condition clause. You need to escape PHQL reserved words using [ and ] delimiters
 *
 *<code>
 *	$builder->having('SUM(Robots.price) > 0');
 *</code>
 *
 * @param string $having
 * @return Phalcon\Db\Builder\Select
 */
PHP_METHOD(Phalcon_Db_Builder_Select, having){

	zval *having;

	phalcon_fetch_params(0, 1, 0, &having);

	phalcon_update_property_array_str(getThis(), SL("_definition"), SL("having"), having);
	RETURN_THIS();
}

/**
 * Sets a LIMIT clause, optionally a offset clause
 *
 * @param int $limit
 * @param int $offset
 * @return Phalcon\Db\Builder\Select
 */
PHP_METHOD(Phalcon_Db_Builder_Select, limit){

	zval *limit, *offset = NULL;

	phalcon_fetch_params(0, 1, 1, &limit, &offset);

	phalcon_update_property_array_str(getThis(), SL("_definition"), SL("limit"), limit);

	if (offset) {
		phalcon_update_property_array_str(getThis(), SL("_definition"), SL("offset"), offset);
	}

	RETURN_THIS();
}

/**
 * Sets an OFFSET clause
 *
 * @param int $offset
 * @return Phalcon\Db\Builder\Select
 */
PHP_METHOD(Phalcon_Db_Builder_Select, offset){

	zval *offset;

	phalcon_fetch_params(0, 1, 0, &offset);

	phalcon_update_property_array_str(getThis(), SL("_definition"), SL("offset"), offset);
	RETURN_THIS();
}

/**
 * Sets a GROUP BY clause
 *
 * @param string|array $group
 * @return Phalcon\Db\Builder\Select
 */
PHP_METHOD(Phalcon_Db_Builder_Select, groupBy){

	zval *group;

	phalcon_fetch_params(0, 1, 0, &group);

	phalcon_update_property_array_str(getThis(), SL("_definition"), SL("group"), group);
	RETURN_THIS();
}

/**
 * Returns a PHQL statement built based on the builder parameters
 *
 * @return Phalcon\Db\ResultInterface
 */
PHP_METHOD(Phalcon_Db_Builder_Select, _execute){

	zval *count = NULL, definition = {}, conditions = {}, joins = {}, bind_params = {}, bind_types = {}, service = {}, dependency_injector = {}, connection = {}, dialect = {}, sql_select = {};

	phalcon_fetch_params(1, 0, 1, &count);

	if (!count) {
		count = &PHALCON_GLOBAL(z_false);
	}

	phalcon_read_property(&definition, getThis(), SL("_definition"), PH_SEPARATE);
	PHALCON_MM_ADD_ENTRY(&definition);

	phalcon_read_property(&conditions, getThis(), SL("_conditions"), PH_READONLY);
	if (PHALCON_IS_NOT_EMPTY(&conditions)) {
		phalcon_array_update_str(&definition, SL("where"), &conditions, PH_COPY);
	}

	phalcon_read_property(&joins, getThis(), SL("_joins"), PH_READONLY);
	if (PHALCON_IS_NOT_EMPTY(&joins)) {
		phalcon_array_update_str(&definition, SL("joins"), &joins, PH_COPY);
	}

	PHALCON_MM_CALL_SELF(&bind_params, "getbindparams");
	PHALCON_MM_ADD_ENTRY(&bind_params);

	PHALCON_MM_CALL_SELF(&bind_types, "getbindtypes");
	PHALCON_MM_ADD_ENTRY(&bind_types);

	phalcon_read_property(&service, getThis(), SL("_defaultConnectionService"), PH_READONLY);
	if (Z_TYPE(service) != IS_OBJECT) {
		if (PHALCON_IS_EMPTY(&service)) {
			PHALCON_MM_THROW_EXCEPTION_STR(phalcon_db_builder_exception_ce, "Invalid injected connection service");
			return;
		}

		PHALCON_MM_CALL_METHOD(&dependency_injector, getThis(), "getdi");
		PHALCON_MM_ADD_ENTRY(&dependency_injector);
		if (Z_TYPE(dependency_injector) != IS_OBJECT) {
			PHALCON_MM_THROW_EXCEPTION_STR(phalcon_db_builder_exception_ce, "A dependency injector container is required to obtain the services related to the ORM");
			return;
		}

		/**
		 * Request the connection service from the DI
		 */
		PHALCON_MM_CALL_METHOD(&connection, &dependency_injector, "getshared", &service);
		PHALCON_MM_ADD_ENTRY(&connection);
		if (Z_TYPE(connection) != IS_OBJECT) {
			PHALCON_THROW_EXCEPTION_STR(phalcon_db_builder_exception_ce, "Invalid injected connection service");
			return;
		}
	} else {
		ZVAL_COPY_VALUE(&connection, &service);
		PHALCON_MM_VERIFY_INTERFACE(&connection, phalcon_db_adapterinterface_ce);
	}

	PHALCON_MM_CALL_METHOD(&dialect, &connection, "getdialect");
	PHALCON_MM_ADD_ENTRY(&dialect);

	PHALCON_MM_CALL_METHOD(&sql_select, &dialect, "select", &definition, count);
	PHALCON_MM_ADD_ENTRY(&sql_select);

	/**
	 * Execute the query
	 */
	PHALCON_MM_CALL_METHOD(return_value, &connection, "query", &sql_select, &bind_params, &bind_types);

	RETURN_MM();
}

/**
 * Returns a PHQL statement built based on the builder parameters
 *
 * @return Phalcon\Db\ResultInterface
 */
PHP_METHOD(Phalcon_Db_Builder_Select, count){

	PHALCON_CALL_METHOD(return_value, getThis(), "_execute", &PHALCON_GLOBAL(z_true));
}
